highwind-ai/examples
# Feature/add knn recommender scripts

### Overview:

This PR includes an analysis of common attributes in a banking dataset that can be leveraged for various recommendation systems. Additionally, it utilizes the Santander dataset and implements a KNN model.

- A notebook using the Santander dataset and a simple KNN model to make user banking product recommendations.
- A text file containing the common attributes that appear for a banking dataset based on the AB and Santander datasets.
- Deployment folder which contains resources required for deploying the model onto Highwind.
- Training folder which contains the kfp python containerised template scripts and pipeline logic


### Changes:

- A Notebook that includes dataset analysis, pre-processing, knn model, using the saved model to make recommendations.
- README file that depicts how to run the Notebook
- A .txt file (`common-banking-dataset-fields.txt`) which contains  the common fields for the banking dataset for the recommender model
- A requirements file with the necessary libraries
- Training folder with the following scripts:
     - `data_prep`: Responsible for loading the data, renaming columns, handling data type conversions, dealing with missing values, and updating specific columns.

   - `feature_engineering`: Handles encoding categorical features, scaling features, and preparing the data for modelling.

   -  `model_training`: Conducts training of a KNN model on PCA-transformed training data.

   - `pipeline.py`: Manages KFP pipeline construction logic.
   
   - README file which has usage instructions on building the KFP image component 

### How to test: Notebook and Inference (deployment folder)
- Follow the instructions in the main README and the README in the `deployment folder`
- Ensure you have a HuggingFace access token (The generation of the token is specified in the Notebook) to be able to access the dataset and download the required files.
- If you want to test the saved model to make recommendations from the Notebook, you just need to go to the cell called `Get recommendations using the saved model` and run all proceeding cells.
     - Ensure you add your access token and run the imports and config cells first.

### How to test: Training folder (kfp)
- **NB:** The train folder testing is optional at this stage, as KFP is still under development from the engineering side. On the ML side, we aimed to use one of the recommender examples to understand how to compile a KFP pipeline and integrate the logic into the required kfp python containerized template. 
 - To test that you can successfully build and compile the kfp pipeline, follow the instructions in the README within the `training/README.md` directory.
 - You will be required to generate your target image for the KFP build command to work. 
 - Once the build is complete, you should see the following kfp generated files in green:
   
<img width="335" alt="image" src="https://github.com/user-attachments/assets/988ee585-6656-4668-b43c-fa6f91dfbed0">
 
### Additional Information:

- The Santander dataset and associated model files are currently placed on HuggingFace in the `Melio AI organization`. It's currently private (`dataset`) but should be moved to a public `model` card.

- The Santander dataset doesn't have a target variable like the AB data, thus the KNN model needed to be updated. 

### Potential Issues:
 - You might run into the following error when running the KFP build command `Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))`

- If you’re using Docker Desktop, go to Settings > Advanced and check `Allow the default Docker socket to be used (requires password)`. This should resolve the issue. You might need to restart Docker for the changes to take effect.
